### DISCUSSION TITLE
Understanding latency spikes during evalSha calls

### DISCUSSION THREAD

**Discussion by AdithyaViswanathan** [redisson/redisson#6174] [Q&A]
I am encountering intermittent latency spikes of up to 500ms in my redisson application when using the `evalSha` API to execute scripts on the master node.

**Script:**
My script is simple, I iterate once or twice depending on the key size (mostly 2), and perform SCARD, SADD to them. So totally around 4 commands per script. Usually the script takes < 100 microseconds to complete.

**Redisson configuration**
```
redissonVersion : 3.25.0
masterConnectionPoolSize: 64
masterConnectionMinimumIdleSize: 64
nettyThreads: 0 (32 threads)
slaveConnectionPoolSize:3
slaveConnectionMinimumIdleSize:3
```

**Timing the evalSha call**
```
long startTime = System.nanoTime();

List<String>returnVal = rScript.evalSha(
        RScript.Mode.READ_WRITE,
        scriptSha,
        RScript.ReturnType.MULTI,
        keys,
        arguments);

long elapsedTime = System.nanoTime() - startTime;
```

I see an increase in the latency for certain requests. And I see few patterns on where the time is spent using trace logs..

1.  **Connection Release latency:** I've noticed added latency when a connection is released by a redisson-netty thread
```
2024-09-11 15:40:52.723 TRACE client.handler.CommandDecoder:141 - [redisson-netty-2-12] 
2024-09-11 15:40:52.787 DEBUG redisson.command.RedisExecutor:668 - [redisson-netty-2-12] connection released for command (EVALSHA) and params
```

2. **Encoding Time:** The encoding takes longer than expected occasionally. I use
StringCodec, and I suspect if the context switch to a redisson-netty thread took extra time.

```
2024-09-12 13:39:34.238 DEBUG redisson.command.RedisExecutor:633 - [parallel-10] - acquired connection for command
2024-09-12 13:39:34.320 TRACE client.handler.CommandEncoder:133 - [redisson-netty-2-20] - message: *9
```

3. **Connection Acquisition:** Acquiring a connection is slow sometimes, even with a pool of 64 connections to the master.

```
2024-09-12 16:36:30.715 INFO Application.RedisCaller:58 - [parallel-15] - Calling Redis using Script
2024-09-12 16:36:30.782 DEBUG redisson.command.RedisExecutor:633 - [parallel-15] -  acquired connection for command (EVALSHA) and params
```
 
4. **Redis call latency:**  I see 100ms latency between the encoder and decoder log lines. The redis instance is processing the script faster, as I see multiple redis script calls to the same master, performed within the 100ms from other application we host. So, i am not sure if the network was slow here, or the redisson-netty decoder was slow.

```
2024-09-12 13:50:12.102 TRACE client.handler.CommandEncoder:133 - [redisson-netty-2-8] - channel: [id: 0xd0c1c61c, L:/10.32.28.203:39320 - R:10.126.68.164/10.126.68.164:30181] message: *9
2024-09-12 13:50:12.217 TRACE client.handler.CommandDecoder:141 - [redisson-netty-2-8] - reply: 
```

**Questions**:
Are there any specific configuration settings or best practices that could help optimize the performance?
How should I debug such an issue?, where should i be looking for errors?

Appreciate any help, and please let me know if you need any detailed information from my log.

**Comment by mrniko** [comment]
> and I suspect if the context switch to a redisson-netty thread took extra time.

Encoding is executed on the call thread and not redisson-netty thread

**Reply by mrniko** [reply]
The issue is with the resources on Redis and Redisson side and maybe network

**Reply by AdithyaViswanathan** [reply]
@mrniko 
Thanks for the reply. 

1. If the encoding is on the call thread, I see the redisson-netty thread printing the log from CommandEncoder. And if it was any problem with the encoder, I suspect every request to be having high latency , but some requests finish in <1ms. That's why I suspected on context switch. 
2. Can you elaborate on what you mean by issue on redis and redisson side. I tried running `SLOW-LOG` and `LATENCY DOCTOR` on redis side, and the command execution was fast. So should I add more netty threads? More connections to master?  

How should I further debug this issue? Where can the issue be?

Appreciate any help. Thanks

**Reply by AdithyaViswanathan** [reply]
Could the default ping Connection interval be a problem? maybe all redisson-netty threads were busy with the PING commands? But it's every 30 seconds though.

**Reply by AdithyaViswanathan** [reply]
Hello, just to understand better, can you give me a situation when I can experience increased latency for the above 4 scenarios. I am not sure when does connection release take more time? Or at what scenarios can situation 2 happen?
There aren't enough documentation around this topic, and would be happy if you could provide some guidance

**Comment by AdithyaViswanathan** [comment]
@mrniko May I know why does redisson - netty thread pick up the job, when I use `evalSha`. If the calling thread is going to get blocked anyways, why do we switch to a redisson netty thread. Is there any way where I can turn this context switch off?
